lesson-8-Character-Devices/mpu6050.c

```diff
@@ -0,0 +1,285 @@
+	printk(KERN_INFO "mpu6050: i2c driver removed\n");
+	return 0;
+}
+
+static const struct i2c_device_id mpu6050_idtable[] =
+		{ { "gl_mpu6050", 0 }, { } };
+
+MODULE_DEVICE_TABLE( i2c, mpu6050_idtable);
+
+static struct i2c_driver mpu6050_i2c_driver = { .driver =
+		{ .name = "gl_mpu6050", },
+
+.probe = mpu6050_probe, .remove = mpu6050_remove, .id_table = mpu6050_idtable, };
+
+static int major = 0;
+module_param( major, int, S_IRUGO);
+
+static ssize_t dev_read(struct file * file, char * buf, size_t count,
+		loff_t *ppos) {
+	char b[100];
+	int len = 0;
+	int last = number - 1;
+	if (last < 0) {
+		last = BUF_SIZE - 1;
+	}
+	sprintf(b,
+			"{\"temp\": \"%d\",\"ax\": \"%d\",\"ay\": \"%d\",\"az\": \"%d\",\"gx\": \"%d\",\"gy\": \"%d\",\"gz\": \"%d\"}\n",
+			g_mpu6050_data[last].temperature,
+			g_mpu6050_data[last].accel_values[0],
+			g_mpu6050_data[last].accel_values[1],
+			g_mpu6050_data[last].accel_values[2],
+			g_mpu6050_data[last].gyro_values[0],
+			g_mpu6050_data[last].gyro_values[1],
+			g_mpu6050_data[last].gyro_values[2]);
+
+	len = strlen(b);
+	printk( KERN_INFO "=== read : %ld\n", (long)count );
+	printk( KERN_INFO "=== len : %ld\n", (long)len );
+	if (count < len)
+		return -EINVAL;
+	if (*ppos != 0) {
+		printk( KERN_INFO "=== read return : 0\n" );  // EOF
+		return 0;
+	}
+	if (copy_to_user(buf, b, len))
+		return -EINVAL;
+	*ppos = len;
+	printk( KERN_INFO "=== read return : %d\n", len );
+	return len;
+}
+
+static ssize_t dev_all(struct file * file, char * buf, size_t count,
+		loff_t *ppos) {
+	char b[100];
+	int len = 0;
+	int i = 0;
+	char string[1000] = "";
+	for (; i < BUF_SIZE; i++) {
+		sprintf(b, "%llu:gyro=%d,%d,%d acc=%d,%d,%d\n",
+				g_mpu6050_data[i].timestamp, g_mpu6050_data[i].gyro_values[0],
+				g_mpu6050_data[i].gyro_values[1],
+				g_mpu6050_data[i].gyro_values[2],
+				g_mpu6050_data[i].accel_values[0],
+				g_mpu6050_data[i].accel_values[1],
+				g_mpu6050_data[i].accel_values[2]);
+		strcat(string, b);
+	}
+
+	len = strlen(string);
+	printk( KERN_INFO "=== read : %ld\n", (long)count );
+	printk( KERN_INFO "=== len : %ld\n", (long)len );
+	if (count < len)
+		return -EINVAL;
+	if (*ppos != 0) {
+		printk( KERN_INFO "=== read return : 0\n" );  // EOF
+		return 0;
+	}
+	if (copy_to_user(buf, string, len))
+		return -EINVAL;
+	*ppos = len;
+	printk( KERN_INFO "=== read return : %d\n", len );
+	return len;
+}
+
+static const struct file_operations dev_fops = { .owner = THIS_MODULE, .read =
+		dev_all, };
+
+static const struct file_operations dev_fops1 = { .owner = THIS_MODULE, .read =
+		dev_read, };
+
+#define DEVICE_FIRST  0
+#define DEVICE_COUNT  2
+#define MODNAME "mpu6050_mod"
+#define DEVNAME "dyn"
+
+static struct cdev hcdev;
+static struct cdev hcdev1;
+static struct class *devclass;
+
+int thread_func(void *a) {
+
+	while (true) {
+		if (++number >= BUF_SIZE) {
+			number = 0;
+		}
+		ssleep(1);
+		if (kthread_should_stop()) {
+			do_exit(0);
+		}
+		mpu6050_read_data(&g_mpu6050_data[number]);
```

Comment:
This's not a big deal, but, AFAIR, the task was to read sensor data only by request (but not more often than once per second) - not each second.

But there's an actual issue - no synchronization between reading and writing `number` and `g_mpu6050_data[]`.